### PR TITLE
Adding uri restriction api in shoebox, Adding adjusting score weights in admin, Adding generic json formatting for enumeration type, + unit test

### PR DESCRIPTION
**Pull request by lawzlo**
@stkem @yingjieMiao @leogrim @lawzlo

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala [hunk 1]
```diff
@@ -341,6 +343,7 @@ object Graph extends Service {
 
 object Curator extends Service {
   object internal {
+    //scoreCoefficient key should be one of "socialScore", "popularityScore", "overallInterestScore", "recentInterestScore", "recencyScore", "priorScore"
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala]: I don't see a `scoreCoefficient`, am I missing something?

File: kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala [hunk 2]
```diff
@@ -341,6 +343,7 @@ object Graph extends Service {
...
     def adHocRecos(userId: Id[User], n: Int) = ServiceRoute(GET, "/internal/curator/adHocRecos", Param("userId", userId), Param("n", n))
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala]: You probably want to change it to a POST request

File: kifi-backend/modules/common/app/com/keepit/model/ScoreType.scala [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package com.keepit.model
+
+import com.keepit.common.util.EnumFormatUtil
+import play.api.libs.json._
+
+object ScoreType extends Enumeration {
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/model/ScoreType.scala:6]: Oh, I didn't know that was a thing. Nice!

File: kifi-backend/modules/common/app/com/keepit/model/ScoreType.scala [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  type ScoreType = Value
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/model/ScoreType.scala:7]: Where does this pattern come from? I have to admit I don't quite understand how the serialization works. Perhaps you can walk me through it tomorrow.

File: kifi-backend/modules/common/test/com/keepit/shoebox/FakeShoeboxServiceClientImpl.scala [hunk 5]
```diff
@@ -597,6 +597,8 @@ class FakeShoeboxServiceClientImpl(val airbrakeNotifier: AirbrakeNotifier) exten
 
   def getURISummaries(uriIds: Seq[Id[NormalizedURI]]): Future[Map[Id[NormalizedURI], URISummary]] = Future.successful(Map.empty)
 
+  def getAdultRestrictionOfURIs(uris: Seq[Id[NormalizedURI]]): Future[Seq[Boolean]] = Future.successful(Seq.empty)
```
Comment (stkem) [inline kifi-backend/modules/common/test/com/keepit/shoebox/FakeShoeboxServiceClientImpl.scala]: The returned `Seq` should match the length of the input since that is the contract on the real method. It can just be all `false`.

File: kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala [hunk 6]
```diff
@@ -14,17 +16,32 @@ import com.google.inject.{ Inject, Singleton }
 @Singleton
 class RecommendationGenerationCommander @Inject() (
     seedCommander: SeedIngestionCommander,
+    shoebox: ShoeboxServiceClient,
     scoringHelper: UriScoringHelper) {
 
-  def getAdHocRecommendations(userId: Id[User], howManyMax: Int): Future[Seq[Recommendation]] = {
-    val seedsFuture = seedCommander.getTopItems(userId, Math.max(howManyMax, 150))
+  val defaultScore = 0.0f
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala:22]: After some thought, I think we should make the default scoring a little bit smarter. Let's go over that together.

File: kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala [hunk 7]
```diff
@@ -14,17 +16,32 @@ import com.google.inject.{ Inject, Singleton }
...
           userId = scoredItem.userId,
           uriId = scoredItem.uriId,
-          score = 0.5f * scoredItem.uriScores.recencyScore + 2 * scoredItem.uriScores.overallInterestScore + 0.25f * scoredItem.uriScores.priorScore, //this math is just for testing
+          score = scoreCoefficients.getOrElse(ScoreType.recencyScore, defaultScore) * scoredItem.uriScores.recencyScore
+            + scoreCoefficients.getOrElse(ScoreType.overallInterestScore, defaultScore) * scoredItem.uriScores.overallInterestScore
+            + scoreCoefficients.getOrElse(ScoreType.priorScore, defaultScore) * scoredItem.uriScores.priorScore
+            + scoreCoefficients.getOrElse(ScoreType.socialScore, defaultScore) * scoredItem.uriScores.socialScore
+            + scoreCoefficients.getOrElse(ScoreType.popularityScore, defaultScore) * scoredItem.uriScores.popularityScore
+            + scoreCoefficients.getOrElse(ScoreType.recentInterestScore, defaultScore) * scoredItem.uriScores.recentInterestScore,
+          //this math is just for testing
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala]: this comment is no longer applicable

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/RecommendationsController.scala [hunk 8]
```diff
@@ -19,7 +21,11 @@ class RecommendationsController @Inject() (
 
   def adHocRecos(n: Int) = JsonAction.authenticatedAsync { request =>
```
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/RecommendationsController.scala:22]: We have a `JsonToJsonAction` which will make sure the body of the request is actually json (i.e. no need for `.asJson`) and provides meaningful errors to the client if it is not. Makes your life a little easier.

File: kifi-backend/modules/common/app/com/keepit/common/util/MapFormatUtil.scala [hunk 9]
```diff
@@ -0,0 +1,23 @@
+package com.keepit.common.util
+
+import com.keepit.model.ScoreType
+import com.keepit.model.ScoreType._
+import play.api.libs.json._
+
+object MapFormatUtil {
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/util/MapFormatUtil.scala:9]: Please add a short test for both of the format utils

**Comment by stkem** [comment]
nice work! next step (after comments) is to get your branches sorted out :-)

File: kifi-backend/modules/common/app/com/keepit/common/util/MapFormatUtil.scala (hunk 9 above)
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/util/MapFormatUtil.scala:9]: On second thought, forget about this for now. We'll do it in a separate pull request.

File: kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala (hunk 2 above)
Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala]: right. fixing

File: kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala (hunk 1 above)
Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala]: Deleting comment.

File: kifi-backend/modules/common/test/com/keepit/shoebox/FakeShoeboxServiceClientImpl.scala (hunk 5 above)
Comment (lawzlo) [inline kifi-backend/modules/common/test/com/keepit/shoebox/FakeShoeboxServiceClientImpl.scala]: fixing

File: kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala (hunk 6 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala:22]: Cool.

File: kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala (hunk 7 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala]: Deleting comment

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/RecommendationsController.scala (hunk 8 above)
Comment (lawzlo) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/RecommendationsController.scala:22]: I checked code base, seems JsonToJsonAction be commented out and not be used any more? weird.

File: kifi-backend/modules/common/app/com/keepit/model/ScoreType.scala (hunk 4 above)
Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/model/ScoreType.scala:7]: I think this is a typical pattern of defining Enumeration type.

Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/model/ScoreType.scala:7]: Here is a good explanation of using Enumeration: http://stackoverflow.com/questions/11067396/understanding-scala-enumerations

File: kifi-backend/modules/common/app/com/keepit/model/ScoreType.scala (hunk 3 above)
Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/model/ScoreType.scala:6]: Actually, I was not familiar with Enumeration before but still spent some time to create one for typing scores instead of using String. I think it's necessary because we may have more scores in the future.  I heard Facebook Feeds have more than 1000 features(scores) to generate a feed. I think they have a class/file to manage those features for easy add and delete. :)

File: kifi-backend/modules/common/app/com/keepit/common/util/MapFormatUtil.scala (hunk 9 above)
Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/common/util/MapFormatUtil.scala:9]: Ok.

**Comment by lawzlo** [comment]
Complied, tested. Pushing to Prod.

**Comment by stkem** [comment]
:+1:
